### ISSUE TITLE
Architecture for including pending posts on the homepage

### ISSUE THREAD

**Issue by vsinghkpcc** [SCPR/SCPRv4#868]
AC

- [ ] There is documentation around how we would include pending posts on the homepage for scheduled publishing

**Comment by Ravenstine** [comment]
Here's the scoop so far:

The backend implementation for pending posts seems fairly easy.  A piece of content(i.e. HomepageContent) can have an optional timestamp at which it will become visible, and we could have a scheduled task to periodically refresh the homepage cache if any of the homepage contents have become visible.  This has to happen because the actual homepage records are not being queried on page load; rather, the whole page is retrieved from a cache.  I'd prefer not to diverge from this as it does allow our homepage to load very fast, and the experience of speedy loading should take precedence over pending content.

The bigger issue is on the frontend – the Outpost aggregator is not really designed to be an editing interface for individual pieces of content.  It could potentially be made to support that, but then we'd need to add special exceptions for homepages.  There is a second interface I built a long time ago for modifying the display scheme of homepage content, though I'd advise that the time spent adding more features to it would be better spent making a better homepage editor in general that doesn't depend on things like the aggregator(which should just do one thing).  One of the chief reasons for this is that the quickest win to support display schemes was to write a display editor that could integrate with the aggregator, which is built on Backbone.  I made some decisions at the time that I think no longer make sense – the biggest one being [framework.js.coffee](https://github.com/SCPR/SCPRv4/blob/master/app/assets/javascripts/framework.js.coffee), which extends Backbone to make it less tedious to use.  At this point, we shouldn't hesitate to build more specialized tools for these kind of situations.  Perhaps those of us with expertise in libraries like React can make a fancier homepage editor for Outpost that ties in to the same API the aggregator uses?

An alternative would be to have different versions of the homepage(as opposed to the one we keep editing) that are set to publish at different times.  So someone could "clone" the existing homepage, add a piece of content, and set it to publish those changes at "6pm".  We would still need a scheduled task to check if a version of the homepage has become current.  The only downside I can see from this is that you couldn't set multiple pieces of content to publish at different times, but I don't think we really need control that granular.

**Comment by nakedsushi** [comment]
> we could have a scheduled task to periodically refresh the homepage cache if any of the homepage contents have become visible

How would existing published content play into this? Would the new `scheduled_at` timestamp be null for them, or would we run a migration to populate their `scheduled_at` attributes to be sometime in the past?

@vsinghkpcc can there ever be a time where a post has already been published in say, a show page or vertical landing, but just hasn't been published on the homepage?  Also, do we have rough numbers of % of content that at one point or another ends up on the homepage?  I'm concerned about adding a column to all content if only a small % of them actually need it.

I like the idea of having different versions of the homepage that are published at different times, since that's closer to the problem we're trying to solve.  We could have the scheduled task update the homepage every half-hour -- would that be granular enough? Then in theory we'd have at most 48 versions of a homepage in a day if needed.  Users would also be able to just go back and update the most recently published homepage if they need to change something on the fly.

**Comment by Ravenstine** [comment]
> How would existing published content play into this? Would the new scheduled_at timestamp be null for them, or would we run a migration to populate their scheduled_at attributes to be sometime in the past?

@nakedsushi I would expect a nil value to indicate no schedule and to publish immediately.  So far we've always published things immediately and I would bet that 95+% of the things we publish would just get published immediately, therefore that's the correct behavior.  Otherwise it would get irritating for someone to add a timestamp to every story they include in a homepage.  Then again, a default timestamp could be added on create(Time.now), but that would literally have the same effect as just considering nil to be "publish asap".

We also don't need to add a column to all content  – a `scheduled_at` attribute would appear in the table for the HomepageContent join model.  So only one table need be affected.

I also agree that it's better to have different versions of the homepage, and that's basically what I originally had in mind; somehow we ended up always editing one homepage record...  All the attributes for that are basically in place for individual homepages.  The key changes to this would be to change the query for how the current homepage is defined(e.g. `where published_at is in the past`), add a "clone" button to the homepage editor, and have a scheduled task periodically refresh the homepage cache.

**Comment by Ravenstine** [comment]
So to finalize this, here is a rough outline of what would need to change:

- The `current` scope for BetterHomepages should only return a homepage where `published_at` is in the past, not the future.
- Add a task to `lib/tasks/scheduler.rake` that periodically runs `Job::BetterHomepageCache.enqueue`.  The period for each run is TBD.
- Add a "clone" button to the homepage editor to create a new duplicate of an existing homepage.

If there's anything else that needs to be discussed, feel free to reopen this ticket.  I feel like this is the best approach because we don't really need homepages to be first class citizens, meaning a short delay between periodic refreshes shouldn't be a problem for us especially when they are predictable.  Having individual pieces of content have their own timestamps triggering their respective homepages to refresh is more complicated and can lead to more confusion("why isn't X showing up on the homepage?").

**Comment by nakedsushi** [comment]
Just to clarify, `current` scope for BetterHomepages should only return the latest homepage where `published_at` is in the past, right?

I like this approach. It may be worth running it by Brian Frank or Mike Roe to see if it makes sense to them. Considering it's essentially, different drafts of the homepage to be published at a scheduled date, I *think* it'll be easy for them to set up editorially.

**Comment by vsinghkpcc** [comment]
This is all generally in line with what I've talked to Brian about so far, so I think we're headed in the right direction.

and @nakedsushi yes, there are times a story is published somewhere else on the site (ex. a show segment that appears on a show page) and doesn't ever / takes a few hours for it to appear on the homepage. Not sure about the % of stories that get to the homepage, we'd have to look at historical posts. I'd guess it to be something like 70%.

**Comment by Ravenstine** [comment]
Gonna update this ticket due to some prior confusion.

If we want content to be added to the homepage when it automatically moves from `pending` status to `published`, this is what would need to happen:

- Pending content needs to appear in the Outpost aggregator in the homepage editor.  The content already gets indexed but the query to ES is only set to return published items.
- A piece of content made to publish through `ContentAlarm.fire_pending` needs to also check if it is related to the current homepage, and if it is, trigger the homepage to refresh.

I otherwise still like the idea of having multiple editions of the homepage with their own `scheduled_at` attribute.  The other bullet points I made in my previous comment could still apply, but the ones I made here could be executed separately.
